## Listas

Listas são estruturas de dados capazes de armazenar multiplos elementos.

Para a criação de uma lista basta colocar os elementos entre virgulas dentro de colchetes `[]`. Como no exemplo abaixo:

In [1]:
nomes_frutas = ["maçã", "banana", "abacaxi"]

In [2]:
nomes_frutas

['maçã', 'banana', 'abacaxi']

In [3]:
numeros = [2, 13, 17, 47]

In [4]:
numeros

[2, 13, 17, 47]

A lista pode conter elementos de tipos diferentes, conforme o exemplo a seguir:

In [5]:
['lorem ipsum', 150, 1.3, [-1, -2]]

['lorem ipsum', 150, 1.3, [-1, -2]]

In [6]:
vazia = []

In [7]:
vazia

[]

### Exercícios:

1. Crie uma lista com o nome das 3 pessoas mais próximas.

### Índices
Assim como nas _strings_, é possível acessar separadamente cada item de uma lista a partir de seus índices:

In [2]:
lista = [100, 200, 300, 400, 500]

In [9]:
lista[0]  # os índices sempre começam em 0

100

In [10]:
lista[2]

300

In [11]:
lista[4]  # último elemento

500

In [12]:
lista[-1]  # outra maneira de acessar o último elemento

500

Como visto no exemplo anterior, ao utilizar um índice negativo os elementos são acessados de trás pra frente, a partir do final da lista:

In [3]:
lista[-2]  # penúltimo elemento

400

In [4]:
lista[-3]  # terceiro

300

In [5]:
lista[-4] # segundo

200

In [8]:
lista[-5] # primeiro

100

Ou pode-se acessar através de _slices_

In [17]:
lista[2:4]

[300, 400]

In [18]:
lista[:3]

[100, 200, 300]

In [19]:
lista[2:]

[300, 400, 500]

Tentar acessar uma posição inválida de uma lista causa um erro:

In [20]:
lista[10]

IndexError: list index out of range

In [21]:
lista[-10]

IndexError: list index out of range

Podemos avaliar se os elementos estão na lista com a palavra `in`, como no exemplo abaixo:

In [22]:
lista_estranha = ['duas palavras', 42, True, ['batman', 'robin'], -0.84, 'hipófise']

In [23]:
42 in lista_estranha

True

In [24]:
'duas palavras' in lista_estranha

True

In [25]:
'dominó' in lista_estranha

False

In [26]:
'batman' in lista_estranha[3] # note que o elemento com índice 3 também é uma lista

True

É possivel obter o tamanho da lista utilizando o método `len()`

In [27]:
len(lista)

5

In [28]:
len(lista_estranha)

6

In [29]:
len(lista_estranha[3])

2

### Removendo itens da lista

Devido à lista ser uma estrutura mutável, é possivel remover elementos de uma lista utilizando o comando `del`.

In [30]:
lista_estranha

['duas palavras', 42, True, ['batman', 'robin'], -0.84, 'hipófise']

In [31]:
del lista_estranha[2]
lista_estranha

['duas palavras', 42, ['batman', 'robin'], -0.84, 'hipófise']

In [32]:
del lista_estranha[-1] # Remove o último elemento da list
lista_estranha

['duas palavras', 42, ['batman', 'robin'], -0.84]

### Exercícios:

1. Utilizando o 'del' remova todos os elementos da lista criada anteriormente, até ter a lista vazia

### Operação com listas

O operador `+` concatena as listas:

In [33]:
a = [1, 2, 3]
b = [4, 5, 6]

In [34]:
c = a + b

In [35]:
c

[1, 2, 3, 4, 5, 6]

O operador `*` repete a lista dado um número de vezes:

In [36]:
[0] * 3

[0, 0, 0]

In [37]:
[1, 2, 3] * 2

[1, 2, 3, 1, 2, 3]

### Métodos de listas

Existem métodos que permitem alterar a lista como, por exemplo, o método `append` que adiciona um  elemento ao final da lista:

In [1]:
l = ['a', 'b', 'c']
l

['a', 'b', 'c']

In [2]:
l.append('e')

In [3]:
l

['a', 'b', 'c', 'e']

Para inserir numa posição qualquer: list.insert(index, obj)

In [5]:
l.insert(3, 'd')
l

['a', 'b', 'c', 'd', 'd', 'e']

`extend` recebe uma lista como argumento e adiciona todos seus elementos a outra:

In [41]:
l1 = ['a', 'b', 'c']
l2 = ['d', 'e']

In [42]:
l1

['a', 'b', 'c']

In [43]:
l2

['d', 'e']

In [44]:
l1.extend(l2)
l1

['a', 'b', 'c', 'd', 'e']

`l2` não é modificado:

In [45]:
l2

['d', 'e']

O método `sort` ordena os elementos da lista em ordem ascendente:

In [10]:
lista_desordenada = ['b', 'z', 'k', 'a', 'h']
lista_desordenada

['b', 'z', 'k', 'a', 'h']

In [12]:
lista_desordenada.sort()
lista_desordenada # Agora está ordenada!

['a', 'b', 'h', 'k', 'z']

### Exercícios:

1. Dado uma lista de números, faça com que os números sejam ordenados e depois inverta a ordem da lista usando slicing

## OBS 1:
É possível transformatar uma string em número, dado que seja um número, por exemplo:

In [14]:
numero = int("2")
numero

2

## OBS 2:
Assim como também é possível a volta:

In [21]:
numero_string = str(1900)
numero_string

'1900'

In [20]:
type(numero_string)

str

## Função range():

Aprendemos a adicionar itens a uma lista, mas, e se fosse necessário produzir uma lista com os números de 1 até 200?

In [23]:
lista_grande = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13] # ???
lista_grande

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]

Em python, existe a função embutida range(), com ela é possível de uma maneira bem simples:

In [25]:
print(list(range(1, 200)))

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199]


Mas além disso, o range pode nos oferecer algumas coisas interessantes, por exemplo, imprimir os números espaçados de 5 em 5, entre 0 e 30:

In [27]:
print(list(range(0, 30, 5)))

[0, 5, 10, 15, 20, 25]


Mas, como na maior parte das vezes, apenas queremos uma lista começando em 0 e indo até o número desejado - 1, a função range() também funciona da seguinte maneira:

In [28]:
print(list(range(10)))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


## Lendo valores do teclado:
Em python também é possível ler do teclado as informações digitadas pelo usuário, e isso é feito por meio da função embutida _input_ da seguinte forma:

In [30]:
valor_lido = input("digite um valor: ")

digite um valor: 10


In [32]:
type(valor_lido) # deve-se notar que o valor lido é SEMPRE do tipo string

str

In [34]:
valor_lido + 10 # ou seja, antes de trabalhar com esse valor, é preciso converter para o tipo correto

TypeError: must be str, not int

Isso pode ser feito usando os operadores _int()_ e _float()_, dependendo do valor esperado:

In [35]:
valor_lido = int(input("digite um valor inteiro: "))

digite um valor inteiro: 10


In [37]:
type(valor_lido)

int

In [38]:
valor_lido + 10

20

In [40]:
valor_lido = float(input("digite um valor decimal: "))

digite um valor decimal: 1.5


In [42]:
valor_lido - 1

0.5

### Exercícios:

1. Dado: my_list = range(1, 11), use slicing para printar apenas os números ímpares.  

# Condicionais

O tipo de dado booleano (bool) refere-se a uma unidade lógica sobre a qual podemos realizar operações, particularmente úteis para o controle de fluxo de um programa. 

A unidade booleana assume apenas 2 valores: Verdadeiro e Falso.

OBS: Essa estrutura binária é a forma com a qual opera o computador (0 e 1).

In [1]:
True

True

In [3]:
type(False)

bool

Qualquer expressão lógica retornará um valor em bool.

In [8]:
2 < 3

True

In [7]:
2 == 5

False

Os operadores lógicos utilizados em programação são:

- ">", maior a, por exemplo 5 > 3

- "<", menor a

- ">=", maior ou igual a

- "<=", menor ou igual a

- "==", igual a

- "!=", diferente de

## Para realizar operações com expressões lógicas, existem:

- "and", e, ele opera segundo a seguinte tabela:

| Valor1     | Valor2     | Resultado  |
|------------|------------|------------|
| Verdadeiro | Verdadeiro | Verdadeiro |
| Verdadeiro | Falso      | Falso      |
| Falso      | Verdadeiro | Falso      |
| Falso      | Falso      | Falso      |

- "or", ou:

| Valor1     | Valor2     | Resultado  |
|------------|------------|------------|
| Verdadeiro | Verdadeiro | Verdadeiro |
| Verdadeiro | Falso      | Verdadeiro |
| Falso      | Verdadeiro | Verdadeiro |
| Falso      | Falso      | Falso      |

- "not", não:

| Valor      | Resultado  |
|------------|------------|
| Verdadeiro | Falso      |
| Falso      | Verdadeiro |

In [17]:
10 > 3 and 2 == 4

False

In [18]:
10 > 3 or 2 == 4

True

In [19]:
not not not 1 == 1

False

Porém, assim como os operadores aritiméticos, os operadores booleanos também possuem uma ordem de prioridade:

_not_ tem maior prioridade que _and_ que tem maior que _or_

In [7]:
not False and True or False

True

# Estruturas de controle

As estruturas de controle servem para decidir quais blocos de código serão executados.

### Exemplo:

_Se_ estiver nublado:

    Levarei guarda-chuva
   
_Senão_:

    Não levarei

## OBS:
Na linguagem python a identação (espaço dado antes de uma linha) é utilizada para demarcar os blocos de código, e são obrigatórios quando se usa estruturas de controle

In [11]:
a = 7
if a > 3:
    print("estou no if")
else:
    print("cai no else")

estou no if


In [15]:
valor_entrada = 10
if valor_entrada == 1:
    print("a entrada era 1")
elif valor_entrada == 2:
    print("a entrada era 2")
elif valor_entrada == 3:
    print("a entrada era 3")
elif valor_entrada == 4:
    print("a entrada era 4")
else:
    print("o valor de entrada não era esperado em nenhum if")

o valor de entrada não era esperado em nenhum if


# Exercícios

1. Escreva um programa que, dados 2 números diferentes (a e b), encontre o menor deles.

2. Para doar sangue é necessário ter entre 18 e 67 anos. Faça um programa que pergunte a idade de uma pessoa e diga se ela pode doar sangue ou não.

3. Considere uma equação do segundo grau f(x) = a \* x² + b \* x + c. A partir dos coeficientes, determine se a equação possui duas raízes reais, uma, ou se não possui. 

 (Dica: delta = b² - 4 \* a \* c, se delta é maior que 0, possui duas raízes reais, se delta é 0, possui uma raiz, e caso delta seja menor que 0 não possui raiz real)
4. Leia  dois  números  e  efetue  a  adição. Caso o valor somado seja maior que 20, este deverá ser apresentado somando-se a ele mais  8;  caso  o  valor  somado  seja  menor  ou  igual  a  20,  este  deverá  ser  apresentado subtraindo-se 5.

5. Leia um número e imprima a raiz quadrada do número caso ele seja positivo ou igual a zero e o quadrado do número caso ele seja negativo.

6. Leia  um  número  inteiro  entre  1  e  12  e escrever o mês correspondente. Caso o usuário digite um número fora desse intervalo, deverá aparecer uma mensagem informando que não existe mês com este número. 


Desafios:
1. Escreva um programa que, dados 3 números diferentes (a, b e c), encontre o menor deles.
2. Dado 3 valores inteiros lidos do teclado: A,  B e C, retorne a soma deles. Porém, caso algum desses valores seja 13, então ele não conta para a soma, e os valores a sua direita também não.

 Por exemplo:

 1, 2, 3 -> 6

 1, 2, 13 -> 3

 1, 13, 3 -> 1

 13, 2, 3 -> 0



# Estruturas de repetição

As estruturas de repetição são utilizadas quando queremos que um bloco de código seja executado várias vezes.

### Em python existem duas formas de criar uma estrutura de repetição:
- O for é usado quando se quer iterar sobre um bloco de código um número determinado de vezes.

- O while é usando quado queremos que o bloco de código seja repetido até que uma condição seja satisfeita, ou seja, é necessário que uma expressão booleana dada seja verdadeira e assim que ela se tornar falsa, o while para.

## OBS:
Na linguagem python a identação é obrigatória, assim como estruturs de controle, para as estruturas de repetição.

In [7]:
# Aqui repetimos o print 3 vezes
for n in range(0, 3):
    print(n)

0
1
2


In [11]:
# Aqui iniciamos o n em 0, e repetimos o print até que seu valor seja maior ou igual a 3
n = 0
while n < 3:
    print(n)
    n += 1

0
1
2


O loop for em python itera sobre os itens de um conjunto, sendo assim, o _range(0, 3)_ precisa ser um conjunto de elementos, e na verdade ele é:

In [12]:
list(range(0, 3))

[0, 1, 2]

Isso se aplica para _strings_ também:

In [13]:
# Para cada letra na palavra, imprimir a letra
palavra = "casa"
for letra in palavra:
    print(letra)

c
a
s
a


In [5]:
lista = [1, 2, 3, 4, 10]
for numero in lista:
    print(numero**2)

1
4
9
16
100


Para auxiliar as estruturas de repetição, existem dois comandos:
- _break_: É usado para sair de um loop, não importando o estado em que se encontra.


- _continue_: Funciona de maneira parecida com a do break, porém no lugar de encerrar o loop, ele faz com que todo o código que esteja abaixo (porém ainda dentro do loop) seja ignorado e avança para a próxima iteração.

In [8]:
""" Esse código deve rodar até que a palavra "sair" seja digitada. 

* Caso uma palavra com 2 ou menos caracteres seja digitada, um aviso deve ser exibido e 
o loop será executado do início (devido ao continue), pedindo uma nova palavra ao usuário.

* Caso qualquer outra palavra diferente de "sair" seja digitada, um aviso deve ser exibido.

* Por fim, caso a palavra seja "sair", uma mensagem deve ser exibida e o loop deve ser
encerrado (break).
"""
while True:
    string_digitada = input("Digite uma palavra: ")
    if string_digitada.lower() == "sair":
        print("Fim!")
        break
    if len(string_digitada) < 2:
        print("String muito pequena")
        continue
    print("Tente digitar \"sair\"")

Digite uma palavra: oi
Tente digitar "sair"
Digite uma palavra: ?
String muito pequena
Digite uma palavra: sair
Fim!


### Exercícios

1. Ler do teclado uma lista com 5 inteiros e imprimir o menor valor
2. Ler do teclado uma lista com 5 inteiros e imprimir True se a lista estiver ordenada de forma crescente ou False caso contrário.
3. Exiba em ordem decrescente todos os números de 500 até 10;
4. Ler do teclado 10 números e imprima a quantidade de números entre 10 e 50
5. Ler do teclado a idade e o sexo de 10 pessoas, calcule e imprima:

 a) idade média das mulheres

 b) idade média dos homens

 c) idade média do grupo
6. Calcule o somatório dos números de 1 a 100 e imprima o resultado.

## Funções

Função é uma sequência de instruções que executa uma operação de computação. Ao definir uma função, você especifica o nome e a sequência de instruções. Depois, pode utilizar (“chamar”) a função pelo nome.

A ideia é similar à função matemática! Mas funções em uma linguagem de programação não realizam necessariamente apenas cálculos.

Vimos o type, um tipo de função: 

In [ ]:
type(23)

In [ ]:
type('textinho')

Criando uma função simples: 

def NOME_DA_FUNÇÃO( LISTA DE PARÂMETROS):<br>
&nbsp;&nbsp;&nbsp;&nbsp;COMANDOS

Obs1: coloque os dois pontos após definir a função!<br>
Obs2: faça a identação nas linhas abaixo da definição da função!

In [ ]:
def soma():
    print(1+1)

In [ ]:
soma()

In [ ]:
def soma():
    return (1+1)

In [ ]:
soma()

## Qual a diferença entre utilizar print e return aqui em cima?!?

In [ ]:
def imprime_letra():
    print("If you didn't care what happened to me. And I didn't care for you")

In [ ]:
imprime_letra()

In [ ]:
type(imprime_letra)

In [ ]:
def repete_letra():
    imprime_letra()
    imprime_letra()

In [ ]:
repete_letra()

# Funções com argumentos

Queremos somar 3 com um número qualquer que insiro na função. Bora lá: 

In [ ]:
def soma_valor(x):
    return 3 + x

In [ ]:
soma_valor(5)

In [ ]:
z = soma_valor(10)
z

Que sem graça! Quero somar dois números quaisquer!

In [ ]:
def soma_dois_numeros(x, y):
    return x + y

In [ ]:
soma_dois_numeros(7, 4)

Tenho dificuldade com a tabuada do 7! Ajude-me!

In [ ]:
def tabuada_do_7():
    for i in range(11):
        print (7 * i)

In [ ]:
tabuada_do_7()

Mai tá legal isso! Quero a tabuada do 1 ao 10 agora! Bora!

In [ ]:
def tabuadas():
    for i in range(1, 11):
        for j in range(1, 11):
            print("%d * %d = %d" % (i, j, i * j))

In [ ]:
tabuadas()

## Exercícios

1. Faça uma função que calcule a área de um círculo. Insira o raio como argumento.

 **dica**: faça a importação de math e use PI de lá.

2. Crie uma função que receba um valor de temperatura em <i>Fahrenheit</i> e transforme em <i>Celsius</i>.

 Relembrar é viver: 

 $\frac{C}{5} = \frac{F - 32}{9}$

3. Faça uma função que determina se um número é par ou ímpar.

 Use o % para determinar o resto de uma divisão. 
 
 Por exemplo: 3 % 2 = 1 e 4 % 2 = 0

4. Crie uma função que receba 3 valores e calcula as raízes da fórmula de Bháskara.

 **dica**: raiz quadrada é sqrt(), importando math: math.sqrt()

 Faça um teste com bhaskara(1, -4, -5) e o programa deve obter as raízes: (5.0, -1.0)
 
5. Dada a função: y = 5x + 2, determine os valores de y para x entre -10 a +10, onde x é inteiro
6. Escreva uma função chamada has_duplicates que tome uma lista e retorne True se houver algum elemento que apareça mais de uma vez. Ela não deve modificar a lista original.
7. Duas palavras são um “par inverso” se uma for o contrário da outra. Escreva uma função que dado duas palavras, retorne True caso sejam.
8. Escreva uma função que imprime todos os números primos entre 1 e 50
 
 **dica**: um número é primo se ele for divisível apenas por 1 e ele mesmo, use o operador % (resto da divisão) para isso.
 
9. Duas palavras são anagramas se você puder soletrar uma rearranjando as letras da outra. Escreva uma função chamada is_anagram que tome duas strings e retorne True se forem anagramas ou False caso contrário.
10. Escreva uma função que dado um número, calcule o fatorial desse número.
 
 Por exemplo, fatorial de 5 = 5\*4\*3\*2\*1 = 120
11. Crie uma função que aproxima a função matemática seno, utilizando a seguinte equação:

 $\sum_{n=0}^{\infty} \frac{(-1)^n}{2n+1!} \cdot x^{2n+1}$